components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/SetAccountAssociationToLocalUserImpl.java

```diff
@@ -33,29 +44,75 @@
+                    String userIdClaimURI = getUserIdClaimURI(federatedIdpName);
+                    String externalSubject;
+                    if (StringUtils.isNotEmpty(userIdClaimURI)) {
+                        externalSubject = federatedUser.getWrapped().getUserAttributes().entrySet().stream().filter(
+                                userAttribute -> userAttribute.getKey().getRemoteClaim().getClaimUri()
+                                        .equals(userIdClaimURI))
+                                .map(Map.Entry::getValue)
+                                .findFirst()
+                                .orElse(null);
+                    } else {
+                        externalSubject = federatedUser.getWrapped().getAuthenticatedSubjectIdentifier();
+                    }
+                    String externalIdpName = federatedUser.getWrapped().getFederatedIdPName();
+                    if (externalSubject != null && externalIdpName != null) {
+                        associateID(externalIdpName, externalSubject, username, tenantDomain, userStoreDomainName);
+                    } else {
+                        if (log.isDebugEnabled()) {
+                            log.debug(" Authenticated user or External IDP may be null " + " Authenticated User: " +
+                                    externalSubject + " and the External IDP name: " + externalIdpName);
+                        }
+                    }
                 } else {
                     if (log.isDebugEnabled()) {
-                        log.debug(" Authenticated user or External IDP may be null " + " Authenticated User: " +
-                                externalSubject + " and the External IDP name: " + externalIdpName);
+                        log.debug("User " + federatedUser.getWrapped().getAuthenticatedSubjectIdentifier() + " " +
+                                "is not a federated user.");
                     }
                 }
             } else {
                 if (log.isDebugEnabled()) {
-                    log.debug("User " + federatedUser.getWrapped().getAuthenticatedSubjectIdentifier() + " " +
-                            "is not a federated user." );
+                    log.debug(" Federated user is null ");
                 }
             }
-        } else {
-            if (log.isDebugEnabled()) {
-                log.debug(" Federated user is null ");
-            }
+        } catch (IdentityProviderManagementException e) {
+            String msg =
+                    "Error while retrieving identity provider by name: " + federatedIdpName;
+            log.error(msg, e);
+        }
+    }
+
+    private String getUserIdClaimURI(String federatedIdpName) throws IdentityProviderManagementException {
+
+        String userIdClaimURI = null;
+        String userNameLocalClaim = "http://wso2.org/claims/username";
+        IdentityProviderManagementService idpManager = new IdentityProviderManagementService();
```

Comment:
If we have an OSGi service to get the IdentityProviderManagementService, let's use it without constructing here.

Reply:
fixed in 4d66091